File: neetcode/blind75/trees/kth-smallest-integer-in-bst.py
# https://neetcode.io/problems/kth-smallest-integer-in-bst

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def kthSmallest(self, root: Optional[TreeNode], k: int) -> int:
        # recursive
        # curr_i = 0
        # val = None

        # # inorder traversal
        # def DFS(root):
        #     nonlocal curr_i
        #     nonlocal val

        #     if not root:
        #         return

        #     left = DFS(root.left)

        #     curr_i += 1
        #     if curr_i == k:
        #         val = root.val
        #         return

        #     right = DFS(root.right)

        # DFS(root)
        # return val

        # iterative
        stack = list()
        curr = root

        while curr or stack:
            while curr:
                stack.append(curr)
                curr = curr.left

            curr = stack.pop(len(stack) - 1)

            k -= 1
            if k == 0:
                return curr.val
                
            curr = curr.right

        return None

File: neetcode/neet150_JS/Trees/kth-smallest-integer-in-bst.js
// https://neetcode.io/problems/kth-smallest-integer-in-bst

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {

    DFS(node, kth) {
        if (node === null) {
            return
        }
        if (kth[0] < 0) {
            return
        }

        this.DFS(node.left, kth)
        kth[0] -= 1
        if (kth[0] === 0) {
            kth[1] = node.val
            return
        }
        this.DFS(node.right, kth)

        return
    }
    /**
     * @param {TreeNode} root
     * @param {number} k
     * @return {number}
     */
    kthSmallest(root, k) {
        if (root === null) {
            return null
        }

        const kth = [k, 0]
        this.DFS(root, kth)
        return kth[1]
    }
}
